pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,399 @@
+	level.Debug(logger).Log("msg", "successfully completed block upload")
+	return nil
+}
+
+// sanitizeMeta sanitizes and validates a metadata.Meta object. If a validation error occurs, an error
+// message gets returned, otherwise an empty string.
+func (c *MultitenantCompactor) sanitizeMeta(logger log.Logger, tenantID string, blockID ulid.ULID,
+	meta *metadata.Meta) string {
+	meta.ULID = blockID
+
+	for l, v := range meta.Thanos.Labels {
+		switch l {
+		// Preserve these labels
+		case mimir_tsdb.CompactorShardIDExternalLabel:
+		// Remove unused labels
+		case mimir_tsdb.DeprecatedTenantIDExternalLabel, mimir_tsdb.DeprecatedIngesterIDExternalLabel, mimir_tsdb.DeprecatedShardIDExternalLabel:
+			level.Debug(logger).Log("msg", fmt.Sprintf("removing unused external label from %s", block.MetaFilename),
+				"label", l, "value", v)
+			delete(meta.Thanos.Labels, l)
+		default:
+			level.Warn(logger).Log("msg", fmt.Sprintf("rejecting unsupported external label in %s", block.MetaFilename),
+				"label", l)
+			return fmt.Sprintf("unsupported external label in %s: %s", block.MetaFilename, l)
+		}
+	}
+
+	for _, f := range meta.Thanos.Files {
+		if f.RelPath == block.MetaFilename {
+			continue
+		}
+
+		if !rePath.MatchString(f.RelPath) {
+			level.Warn(logger).Log("msg", fmt.Sprintf("rejecting file with invalid path in %s", block.MetaFilename),
+				"file", f.RelPath)
+			return fmt.Sprintf("file with invalid path in %s: %s", block.MetaFilename,
+				f.RelPath)
+		}
+
+		if f.SizeBytes <= 0 {
+			level.Warn(logger).Log("msg", fmt.Sprintf("rejecting file with invalid size in %s", block.MetaFilename),
+				"file", f.RelPath)
+			return fmt.Sprintf("file with invalid size in %s: %s", block.MetaFilename,
+				f.RelPath)
+		}
+	}
+
+	// validate minTime/maxTime
+	// basic sanity check
+	if meta.MinTime < 0 || meta.MaxTime < 0 || meta.MaxTime < meta.MinTime {
+		level.Warn(logger).Log("msg", fmt.Sprintf("invalid minTime/maxTime in %s", block.MetaFilename),
+			"minTime", meta.MinTime, "maxTime", meta.MaxTime)
+		return fmt.Sprintf("invalid minTime/maxTime in %s: minTime=%d, maxTime=%d",
+			block.MetaFilename, meta.MinTime, meta.MaxTime)
+	}
+	// validate that times are in the past
+	now := time.Now()
+	if time.UnixMilli(meta.MinTime).After(now) || time.UnixMilli(meta.MaxTime).After(now) {
+		level.Warn(logger).Log("msg", "block time(s) greater than the present", "minTime", meta.MinTime,
+			"maxTime", meta.MaxTime)
+		return fmt.Sprintf("block time(s) greater than the present: minTime=%d, maxTime=%d", meta.MinTime, meta.MaxTime)
+	}
+	// validate data is within the retention period
+	threshold := now.Add(-c.cfgProvider.CompactorBlocksRetentionPeriod(tenantID))
+	if time.UnixMilli(meta.MaxTime).Before(threshold) {
+		maxTimeStr := util.FormatTimeMillis(meta.MaxTime)
+		level.Warn(logger).Log("msg", "block max time older than retention period", "maxTime", maxTimeStr)
+		return fmt.Sprintf("block max time (%s) older than retention period", maxTimeStr)
+	}
```

Comment:
We can only perform this check if `c.cfgProvider.CompactorBlocksRetentionPeriod(tenantID) > 0`. If it's 0, then retention is disabled, and any block should be accepted.

Reply:
@pstibrany thanks!

Reply:
Done.